client/asset/eth/multirpc.go

```diff
@@ -0,0 +1,1270 @@
+		p.ec.Close()
+	}
+	return nil
+}
+
+// cleanReceipts cleans up the receipt cache, deleting any receipts that haven't
+// been access for > receiptCacheExpiration.
+func (m *multiRPCClient) cleanReceipts() {
+	m.receipts.Lock()
+	for txHash, rec := range m.receipts.cache {
+		if time.Since(rec.lastAccess) > receiptCacheExpiration {
+			delete(m.receipts.cache, txHash)
+		}
+	}
+	m.receipts.Unlock()
+}
+
+func (m *multiRPCClient) transactionReceipt(ctx context.Context, txHash common.Hash) (r *types.Receipt, err error) {
+	// Check the cache.
+	const cacheExpiration = time.Minute
+
+	m.receipts.RLock()
+	cached := m.receipts.cache[txHash]
+	if cached != nil && cached.confirmed {
+		cached.lastAccess = time.Now()
+	}
+	if time.Since(m.receipts.lastClean) > time.Minute*20 {
+		m.receipts.lastClean = time.Now()
+		go m.cleanReceipts()
+	}
+	m.receipts.RUnlock()
+
+	// If confirmed or if it was just fetched, return it as is.
+	if cached != nil && (cached.confirmed || time.Since(cached.lastAccess) < cacheExpiration) {
+		return cached.r, nil
+	}
+
+	// Fetch a fresh one.
+	if err = m.withPreferred(func(p *provider) error {
+		r, err = p.ec.TransactionReceipt(ctx, txHash)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	var confs int64
+	if r.BlockNumber != nil {
+		tip, err := m.bestHeader(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("bestHeader error: %v", err)
+		}
+		confs = new(big.Int).Sub(tip.Number, r.BlockNumber).Int64() + 1
+	}
+
+	m.receipts.Lock()
+	m.receipts.cache[txHash] = &receiptRecord{
+		r:          r,
+		lastAccess: time.Now(),
+		confirmed:  confs > txConfsNeededToConfirm,
+	}
+	m.receipts.Unlock()
+
+	return r, nil
+}
+
+type rpcTransaction struct {
+	tx *types.Transaction
+	txExtraDetail
+}
+
+type txExtraDetail struct {
+	BlockNumber *string         `json:"blockNumber,omitempty"`
+	BlockHash   *common.Hash    `json:"blockHash,omitempty"`
+	From        *common.Address `json:"from,omitempty"`
+}
+
+func (tx *rpcTransaction) UnmarshalJSON(b []byte) error {
+	if err := json.Unmarshal(b, &tx.tx); err != nil {
+		return err
+	}
+	return json.Unmarshal(b, &tx.txExtraDetail)
+}
+
+func getRPCTransaction(ctx context.Context, p *provider, txHash common.Hash) (*rpcTransaction, error) {
+	var resp *rpcTransaction
+	err := p.ec.rpc.CallContext(ctx, &resp, "eth_getTransactionByHash", txHash)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil {
+		return nil, asset.CoinNotFoundError
+	}
+	// Just copying geth with this one.
+	if _, r, _ := resp.tx.RawSignatureValues(); r == nil {
+		return nil, fmt.Errorf("server returned transaction without signature")
+	}
+	return resp, nil
+}
+
+func (m *multiRPCClient) getTransaction(ctx context.Context, txHash common.Hash) (tx *types.Transaction, h int64, err error) {
+	return tx, h, m.withPreferred(func(p *provider) error {
+		resp, err := getRPCTransaction(ctx, p, txHash)
+		if err != nil {
+			return err
+		}
+		tx = resp.tx
+		if resp.BlockNumber != nil {
```

Comment:
This `resp.BlockNumber` string has a "0x" in the front, causing the `hex.DecodeString` call to return an error. Also this function should return -1 for the height if the tx has not yet been mined.